src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -867,6 +873,30 @@ public bool StartAsyncOperation(SocketAsyncContext context, TOperation operation
                         return false;
                     }
                 }
+
+                static void HandleFailedRegistration(SocketAsyncContext context, TOperation operation, Interop.Error error)
+                {
+                    Debug.Assert(error != Interop.Error.SUCCESS);
+
+                    // macOS: kevent returns EPIPE when adding pipe fd for which the other end is closed.
+                    if (error == Interop.Error.EPIPE)
+                    {
+                        // Retry the operation.
+                        if (operation.TryComplete(context))
+                        {
+                            return;
+                        }
+                    }
+
+                    if (error == Interop.Error.ENOMEM || error == Interop.Error.ENOSPC)
+                    {
+                        throw new OutOfMemoryException();
+                    }
+                    else
+                    {
+                        throw new InternalException(error);
```

Comment:
I don't think we want InternalException.  Were we previously throwing this type in a similar situation?  Seems like it should be a SocketException or an IOException or something like that.  (I'd actually like to delete InternalException entirely :)

Reply:
We were throwing this exception before. Is `Win32Exception` an option?